### ISSUE TITLE
Change engine version button dosen't work

### ISSUE THREAD

**Issue by Zemogiter** [PhoenicisOrg/phoenicis#1992]
### Expected behavior
A window of some kind should pop out.
### Actual behavior
This is what shows up in terminal when clicking on the button:
```
Exception in thread "pool-3-thread-10" java.lang.UnsupportedOperationException: changeVersion
	at com.oracle.truffle.js.runtime.java.adapter.JavaAdapterServices.unsupported(JavaAdapterServices.java:180)
	at org.phoenicis.engines.Engine$$JSJavaAdapter.changeVersion(Unknown Source)
	at org.phoenicis.javafx.components.container.control.ContainersFeaturePanel.lambda$changeEngineVersion$9(ContainersFeaturePanel.java:169)
	at org.phoenicis.engines.EnginesManager.lambda$getEngine$0(EnginesManager.java:72)
	at org.phoenicis.scripts.session.PhoenicisInteractiveScriptSession.eval(PhoenicisInteractiveScriptSession.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$createInteractiveSession$1(BackgroundScriptInterpreter.java:45)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```
### Steps to reproduce
Go the the Containers tab and click on any container and click on the "change engine version" button
### System information
* software version (git revision) of phoenicis and scripts repositories
https://github.com/PhoenicisOrg/phoenicis/commit/b5ab98e916f70fb3df8db026f2a4b8d76c983e40
script repo https://github.com/PhoenicisOrg/scripts/commit/53a6846f18c9841225d063a1a5d15c7aa94abbdd
* used operating system
Ubuntu 19.04 x64
* used Java version
openjdk 11.0.3 2019-04-16
OpenJDK Runtime Environment (build 11.0.3+7-Ubuntu-1ubuntu219.04.1)
OpenJDK 64-Bit Server VM (build 11.0.3+7-Ubuntu-1ubuntu219.04.1, mixed mode, sharing)

**Comment by ImperatorS79** [comment]
It worked before the change to graalVM.

**Comment by Zemogiter** [comment]
I think we jumped on that "graalvm is better than nashorn" bandwagon too early :(

**Comment by qparis** [comment]
I think it is normal to have some side effect, we are going to fix them. 
Can you provide the step to reproduce? (i.e. what to type in the console)

**Comment by Zemogiter** [comment]
I didn't type anything. Just what I wrote in "Steps to reproduce". I was looking online and found the compability flag and thought we could use it but when I searched in local source directory I already found it.
```
grep -r "nashorn-compat" "./"
Plik binarny ./phoenicis-dist/target/lib/js-scriptengine-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-dist/target/lib/js-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-dist/target/Phoenicis_5.0-SNAPSHOT/usr/share/phoenicis/app/js-scriptengine-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-dist/target/Phoenicis_5.0-SNAPSHOT/usr/share/phoenicis/app/js-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-dist/target/packages/phoenicis/app/js-scriptengine-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-dist/target/packages/phoenicis/app/js-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-scripts/target/classes/org/phoenicis/scripts/engine/ScriptEngineType$2.class pasuje do wzorca
./phoenicis-scripts/src/main/java/org/phoenicis/scripts/engine/ScriptEngineType.java:            return new PolyglotScriptEngine("js", Map.of("js.nashorn-compat", "true"));
```

**Comment by madoar** [comment]
The reason for the error is not GraalVM. The exception is thrown because of a missing changeVersion method in your used engine Implementation. Please check whether the method is contained in the wine implementation in your script repositoryAm 11.06.2019 18:11 schrieb Zemogiter <notifications@github.com>:I didn't type anything. Just what I wrote in "Steps to reproduce". I was looking online and found the compability flag and thought we could use it but when I searched in local source directory I already found it.
grep -r "nashorn-compat" "./"
Plik binarny ./phoenicis-dist/target/lib/js-scriptengine-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-dist/target/lib/js-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-dist/target/Phoenicis_5.0-SNAPSHOT/usr/share/phoenicis/app/js-scriptengine-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-dist/target/Phoenicis_5.0-SNAPSHOT/usr/share/phoenicis/app/js-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-dist/target/packages/phoenicis/app/js-scriptengine-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-dist/target/packages/phoenicis/app/js-19.0.0.jar pasuje do wzorca
Plik binarny ./phoenicis-scripts/target/classes/org/phoenicis/scripts/engine/ScriptEngineType$2.class pasuje do wzorca
./phoenicis-scripts/src/main/java/org/phoenicis/scripts/engine/ScriptEngineType.java:            return new PolyglotScriptEngine("js", Map.of("js.nashorn-compat", "true"));

**Comment by plata** [comment]
Just tried myself and I get:
```
Exception in thread "JavaFX Application Thread" java.lang.IllegalStateException: Multi threaded access requested by thread Thread[JavaFX Application Thread,5,main] but is not allowed for language(s) js.
...
	at org.phoenicis.javafx.views.scriptui.StepRepresentationMenu.drawStepContent(StepRepresentationMenu.java:50)
```

**Comment by plata** [comment]
The problem is in `List<String> menuItems` which is passed to `StepRepresentationMenu`. I'm assuming that we have the issue for any `List` in the wizard. Apparently, the list is a Java object in Javascript or vice versa which may not be accessed.

**Comment by qparis** [comment]
https://medium.com/graalvm/multi-threaded-java-javascript-language-interoperability-in-graalvm-2f19c1f9c37b

**Comment by plata** [comment]
I've seen that but I still don't get it. If I use a simple String, it works.

**Comment by qparis** [comment]
I suspect that it is because this list is watched by a JavaFX thread somewhere else

**Comment by qparis** [comment]
bingo

**Comment by qparis** [comment]
MR is coming

**Comment by plata** [comment]
I can make it work like this:
in `ScriptEngineType`
```Java
public PhoenicisScriptEngine createScriptEngine() {
            return new PolyglotScriptEngine("js",
                    Map.of("js.nashorn-compat", "true", "js.experimental-foreign-object-prototype", "true"));
        }
```
in `changeVersion` (JS Wine implementation)
```
var distributions = new java.util.ArrayList();
        var versions = new java.util.ArrayList();
        wineJson.forEach(function (subPart) {
            var parts = subPart.name.split("-");
            if (parts[2] == architecture) {
                distributions.push(parts[0]);
                versions.push(new java.util.ArrayList());
                subPart.packages.forEach(function (winePackage) {
                    versions[distributions.length-1].push(winePackage.version);
                });
            }
        });
```

**Comment by plata** [comment]
So the problem is the mismatch between the array and the array list.

**Comment by qparis** [comment]
I have a safer patch that do not require modification in JS

**Comment by plata** [comment]
Use `String[]` in Java? Just tried that and it works as well.

**Comment by qparis** [comment]
Should be fixed in #1994

**Comment by qparis** [comment]
The problem is that PolyglotList is directly sent to the UI thread. The solution is to create a copy so that the data becomes immutable once it left the JS thread. Therefore, we are safe

**Comment by plata** [comment]
Actually I tried the same in the beginning. Only one level too deep...

**Comment by qparis** [comment]
Yep, this must be done outside the lambda

**Comment by qparis** [comment]
Can we retest from the master now that #1994 is merged?

**Comment by Zemogiter** [comment]
Downloaded the source code after that pr was merged but now I can't compile it:
```
Running org.phoenicis.javafx.AppConfigurationInitialisationTest
java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204)
	at org.testfx.util.WaitForAsyncUtils$ASyncFXCallable.get(WaitForAsyncUtils.java:715)
	at org.testfx.util.WaitForAsyncUtils.waitFor(WaitForAsyncUtils.java:298)
	at org.testfx.api.FxToolkit.waitForSetup(FxToolkit.java:245)
	at org.testfx.api.FxToolkit.setupFixture(FxToolkit.java:202)
	at org.phoenicis.javafx.AppConfigurationInitialisationTest.testAppConfigurationInitialisation(AppConfigurationInitialisationTest.java:49)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 30.234 s <<< FAILURE! - in org.phoenicis.javafx.AppConfigurationInitialisationTest
[ERROR] testAppConfigurationInitialisation(org.phoenicis.javafx.AppConfigurationInitialisationTest)  Time elapsed: 30.187 s  <<< FAILURE!
java.lang.AssertionError
	at org.phoenicis.javafx.AppConfigurationInitialisationTest.testAppConfigurationInitialisation(AppConfigurationInitialisationTest.java:52)
```

**Comment by qparis** [comment]
Run maven with -DskipTests

**Comment by Zemogiter** [comment]
Ok so after purging the package from the os, compiling with the tests skiped and installing it I confirm the list is working now but the sorting is kinda wonky. It sorts like this:

1. 4.1
2. 4.10
3. 4.2

Reminds me of an solved issue at scripts repo where writing 4.10 in wineVersion would download wine 4.1 instead.

**Comment by qparis** [comment]
Yep, it is because 4.10 was released. I think we have written a version sorter class that can be used. 
So I can close this issue, feel free to open another one if you want
